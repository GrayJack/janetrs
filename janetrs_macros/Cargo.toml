[package]
name = "janetrs_macros"
version = "0.4.6"
authors = ["Eric Shimizu Karbstein <gr41.j4ck@gmail.com>"]
description = "Attribute macros for JanetRS"
repository = "https://github.com/GrayJack/janetrs"
homepage = "https://crates.io/crates/janetrs_macros"
documentation = "https://docs.rs/janetrs_macros"
license-file = "../LICENSE"
edition = "2021"
rust-version = "1.56.0"
keywords = ["janet", "embedding", "proc-macro"]
categories = ["development-tools::procedural-macro-helpers"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
name = "janetrs_macros"
path = "src/lib.rs"
proc-macro = true

[dependencies]
syn = { version = "1", features = ["extra-traits", "full"] }
quote = "1"
proc-macro2 = "1"
janetrs_version = "0.1"

[features]
default = []
system = ["janetrs_version/system"]

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0", features = ["diff"] }
janetrs = { path = "../../janetrs" }
