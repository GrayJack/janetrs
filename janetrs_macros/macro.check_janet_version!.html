<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.check_janet_version.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.check_janet_version.html">macro.check_janet_version.html</a>...</p>
    <script>location.replace("macro.check_janet_version.html" + location.search + location.hash);</script>
</body>
</html>