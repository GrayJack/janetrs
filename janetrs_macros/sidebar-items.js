window.SIDEBAR_ITEMS = {"attr":["cjvg","janet_fn","janet_version"],"macro":["check_janet_version","declare_janet_mod"]};