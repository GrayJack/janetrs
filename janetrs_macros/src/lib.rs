extern crate proc_macro;

use quote::{quote, quote_spanned};
use syn::{parse_macro_input, spanned::Spanned};

#[proc_macro_attribute]
pub fn janet_fn(
    _args: proc_macro::TokenStream, input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let func = parse_macro_input!(input as syn::Item);

    let ts = if let syn::Item::Fn(f) = func {
        // dbg!(&f);
        let f_clone = f.clone();
        let attrs = f.attrs;
        let vis = f.vis;
        let name = f.sig.ident;
        // let syn::ReturnType::Type(_, type_path) = f.sig.output;

        // check for inputs
        if matches!(f.sig.inputs.len(), 1) {
            if let Some(syn::FnArg::Typed(syn::PatType { ty, .. })) = f.sig.inputs.first() {
                if let syn::Type::Reference(syn::TypeReference {
                    mutability,
                    ref elem,
                    ..
                }) = **ty
                {
                    if mutability.is_none() {
                        return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found a immutable reference");}.into();
                    } else if let syn::Type::Slice(syn::TypeSlice {
                        elem: ref slice, ..
                    }) = **elem
                    {
                        if let syn::Type::Path(syn::TypePath { ref path, .. }) = **slice {
                            if !janet_path_checker(path) {
                                return quote_spanned! {path.span() => compile_error!("expected to be a `janetrs::types::Janet`");}.into();
                            }
                        } else {
                            return quote_spanned! {slice.span() => compile_error!("expected to be a `janetrs::types::Janet`");}.into();
                        }
                    } else {
                        return quote_spanned! {elem.span() => compile_error!("expected to be a slice of `janetrs::types::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found something that is not a reference at all");}.into();
                }
            }

            // check output type

            let output_span = match f.sig.output.clone() {
                syn::ReturnType::Default => f_clone.sig.span(),
                syn::ReturnType::Type(_, ty) => ty.span(),
            };
            if let syn::ReturnType::Type(_, ty) = f.sig.output {
                if let syn::Type::Path(syn::TypePath { ref path, .. }) = *ty {
                    if !janet_path_checker(path) {
                        return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
                }
            } else {
                return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
            }

            quote! {
                #(#attrs)*
                #[no_mangle]
                #vis unsafe extern "C" fn #name(argc: i32, argv: *mut ::janetrs::lowlevel::Janet) -> ::janetrs::lowlevel::Janet {
                    #f_clone

                    let args = unsafe { core::slice::from_raw_parts_mut(argv, argc as usize) };
                    let mut args = unsafe { &mut *(args as *mut [::janetrs::lowlevel::Janet] as *mut [::janetrs::types::Janet])};
                    #name(args).into()
                }
            }
        } else {
            quote_spanned! {f_clone.sig.inputs.span() => compile_error!("expected exactly one argument of type `&mut [janetrs::types::Janet]`");}
        }
    } else {
        quote_spanned! {func.span() => compile_error!("expected fn item");}
    };

    ts.into()
}

fn janet_path_checker(path: &syn::Path) -> bool {
    match path.segments.len() {
        1 => {
            let ident = if let Some(i) = path.get_ident() {
                i
            } else {
                return false;
            };
            let test = syn::Ident::new("Janet", ident.span());

            *ident == test
        },
        2 => {
            let mod_types = &path.segments.first().unwrap().ident;
            let janet_ident = &path.segments.last().unwrap().ident;

            let mod_expected = syn::Ident::new("types", mod_types.span());
            let janet_expected = syn::Ident::new("Janet", janet_ident.span());

            *mod_types == mod_expected && *janet_ident == janet_expected
        },
        3 => {
            let mut iter = path.segments.iter();
            let (janetrs_mod, types_mod, janet_ident) = (
                &iter.next().unwrap().ident,
                &iter.next().unwrap().ident,
                &iter.next().unwrap().ident,
            );

            let janetrs_expected = syn::Ident::new("janetrs", janetrs_mod.span());
            let types_expected = syn::Ident::new("types", types_mod.span());
            let janet_expected = syn::Ident::new("Janet", janet_ident.span());

            *janetrs_mod == janetrs_expected
                && *types_mod == types_expected
                && *janet_ident == janet_expected
        },
        _ => false,
    }
}
