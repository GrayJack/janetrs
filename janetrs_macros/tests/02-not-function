use janetrs_macros::*;

#[janet_fn]
struct A(i32);

fn main() {}
