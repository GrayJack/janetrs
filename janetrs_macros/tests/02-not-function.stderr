error: expected fn item
 --> $DIR/02-not-function:4:1
  |
4 | struct A(i32);
  | ^^^^^^^^^^^^^^
